### PR TITLE
Ensure DataParallel replicas can be saved

### LINKED ISSUE #37182: Cannot torch.save replicated modules with v1.5
## 🐛 Bug

With v1.5, `torch.save()` fails on result modules from `torch.nn.parallel.replicate` regardless of `detach=True` or `detach=False`. This error does not occur with v1.4.

## To Reproduce

```python
import torch

r = torch.nn.parallel.replicate(torch.nn.Linear(2, 2).to(0), devices=[0, 1], detach=True)
torch.save(r, "./tmp_module.pt")
```

```python
import torch

r = torch.nn.parallel.replicate(torch.nn.Linear(2, 2).to(0), devices=[0, 1], detach=False)
torch.save(r, "./tmp_module.pt")
```

The above two code snippets both hit the following error. And I verified that the same code works correctly on v1.4.

```
  File "/data/users/shenli/pytorch/torch/serialization.py", line 365, in save
    _legacy_save(obj, opened_file, pickle_module, pickle_protocol)
  File "/data/users/shenli/pytorch/torch/serialization.py", line 430, in _legacy_save
    pickler.dump(obj)
AttributeError: Can't pickle local object 'Module._replicate_for_data_parallel.<locals>.zero_grad'
```

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #37182

The `zero_grad` wrapper from `_replicate_for_data_parallel` can't be pickled. So instead, I set an attribute `_is_replica = True` and check for this in `Module.zero_grad`.

### CODE REVIEW

File: test/distributed/test_data_parallel.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import unittest
 from copy import deepcopy
 from collections import OrderedDict
+import io
```
Comment (mrshenli) [inline test/distributed/test_data_parallel.py:5]: nit: Shall we sort the import? io should be between contextlib and unittest?

File: test/distributed/test_data_parallel.py [hunk 2]
```diff
@@ -669,6 +670,15 @@ def forward(self, input):
         input = torch.randn((8, 8), dtype=torch.float32, device="cuda")
         self.assertTrue(model(input).dtype is torch.float16)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
+    @skipIfRocm
+    def test_save_replica_module(self):
+        # DataParallel replicas can be saved (gh-37182)
+        module = torch.nn.Linear(8, 8).cuda()
+        dpm = torch.nn.parallel.replicate(module, devices=[0, 1])
```
Comment (mrshenli) [inline test/distributed/test_data_parallel.py:678]: shall we cover `detach=True` case as well?

**Review by mrshenli — APPROVED** [review]
Thanks a lot for fixing!! Added two minor comments.

**Review by mrshenli — CHANGES_REQUESTED** [review]
Test failure is real:

```
Apr 25 19:25:16 ======================================================================
Apr 25 19:25:16 FAIL [0.005s]: test_zero_grad (__main__.TestDataParallel)
Apr 25 19:25:16 ----------------------------------------------------------------------
Apr 25 19:25:16 Traceback (most recent call last):
Apr 25 19:25:16   File "distributed/test_data_parallel.py", line 656, in test_zero_grad
Apr 25 19:25:16     dpm(torch.rand(4, 3, 6, 5))
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 562, in __call__
Apr 25 19:25:16     result = self.forward(*input, **kwargs)
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 155, in forward
Apr 25 19:25:16     outputs = self.parallel_apply(replicas, inputs, kwargs)
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 165, in parallel_apply
Apr 25 19:25:16     return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in parallel_apply
Apr 25 19:25:16     output.reraise()
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/_utils.py", line 395, in reraise
Apr 25 19:25:16     raise self.exc_type(msg)
Apr 25 19:25:16 AssertionError: Caught AssertionError in replica 0 on device 0.
Apr 25 19:25:16 Original Traceback (most recent call last):
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
Apr 25 19:25:16     output = module(*input, **kwargs)
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 562, in __call__
Apr 25 19:25:16     result = self.forward(*input, **kwargs)
Apr 25 19:25:16   File "distributed/test_data_parallel.py", line 651, in forward
Apr 25 19:25:16     self.zero_grad()
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/unittest/case.py", line 265, in __exit__
Apr 25 19:25:16     self.expected_regex.pattern, str(first_matching)))
Apr 25 19:25:16   File "/opt/conda/lib/python3.6/unittest/case.py", line 135, in _raiseFailure
Apr 25 19:25:16     raise self.test_case.failureException(msg)
Apr 25 19:25:16 AssertionError: "Calling \.zero_grad\(\) from a module that was passed to a nn\.DataParallel\(\) has no effect." does not match "Calling .zero_grad() from a module created with nn.DataParallel() has no effect. The parameters are copied (in a differentiable manner) from the original module. This means they are not leaf nodes in autograd and so don't accumulate gradients. If you need gradients in your forward method, consider using autograd.grad instead."
```

**Review by mrshenli — APPROVED** [review]

**Comment by v-yunbin** [comment]
> Test failure is real:
> 
> ```
> Apr 25 19:25:16 ======================================================================
> Apr 25 19:25:16 FAIL [0.005s]: test_zero_grad (__main__.TestDataParallel)
> Apr 25 19:25:16 ----------------------------------------------------------------------
> Apr 25 19:25:16 Traceback (most recent call last):
> Apr 25 19:25:16   File "distributed/test_data_parallel.py", line 656, in test_zero_grad
> Apr 25 19:25:16     dpm(torch.rand(4, 3, 6, 5))
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 562, in __call__
> Apr 25 19:25:16     result = self.forward(*input, **kwargs)
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 155, in forward
> Apr 25 19:25:16     outputs = self.parallel_apply(replicas, inputs, kwargs)
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 165, in parallel_apply
> Apr 25 19:25:16     return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in parallel_apply
> Apr 25 19:25:16     output.reraise()
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/_utils.py", line 395, in reraise
> Apr 25 19:25:16     raise self.exc_type(msg)
> Apr 25 19:25:16 AssertionError: Caught AssertionError in replica 0 on device 0.
> Apr 25 19:25:16 Original Traceback (most recent call last):
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 60, in _worker
> Apr 25 19:25:16     output = module(*input, **kwargs)
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 562, in __call__
> Apr 25 19:25:16     result = self.forward(*input, **kwargs)
> Apr 25 19:25:16   File "distributed/test_data_parallel.py", line 651, in forward
> Apr 25 19:25:16     self.zero_grad()
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/unittest/case.py", line 265, in __exit__
> Apr 25 19:25:16     self.expected_regex.pattern, str(first_matching)))
> Apr 25 19:25:16   File "/opt/conda/lib/python3.6/unittest/case.py", line 135, in _raiseFailure
> Apr 25 19:25:16     raise self.test_case.failureException(msg)
> Apr 25 19:25:16 AssertionError: "Calling \.zero_grad\(\) from a module that was passed to a nn\.DataParallel\(\) has no effect." does not match "Calling .zero_grad() from a module created with nn.DataParallel() has no effect. The parameters are copied (in a differentiable manner) from the original module. This means they are not leaf nodes in autograd and so don't accumulate gradients. If you need gradients in your forward method, consider using autograd.grad instead."
> ```

@mrshenli have you solved the problem？

**Pull request closed without merging**